base/abstractarray.jl

```diff
@@ -373,9 +373,16 @@ zero{T}(x::AbstractArray{T}) = fill!(similar(x), zero(T))
 # While the definitions for LinearFast are all simple enough to inline on their
 # own, LinearSlow's CartesianRange is more complicated and requires explicit
 # inlining.
-start(A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
-next(A::AbstractArray,i) = (@_inline_meta(); (idx, s) = next(i[1], i[2]); (A[idx], (i[1], s)))
-done(A::AbstractArray,i) = done(i[1], i[2])
+start(A::AbstractArray) = (@_inline_meta(); _start(linearindexing(A), A))
+_start(::LinearFast, A::AbstractArray) = 1
+_start(::LinearSlow, A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
```

Comment:
You'd never guess it, but this change was motivated by string processing. Specifically, there was a test failure [triggered here](https://github.com/JuliaLang/julia/blob/c18bc5383f6cc4d95cab70e7312a9eb15e0c5b95/test/unicode/utf16.jl#L19), which I fixed (maybe) elsewhere in this PR by [generalizing this function](https://github.com/JuliaLang/julia/blob/c18bc5383f6cc4d95cab70e7312a9eb15e0c5b95/base/unicode/checkstring.jl#L59-L69) and a similar one lower down. That `while pos <= endpos` only works if the iterator is an integer. Since I can understand why people might want to write code like that, I decided to try it and see what bugs surfaced.

Alternatively, we could define comparison operations for `CartesianIndex` iterators and/or change at least the defaults to `start(eachindex(dat))`. That still wouldn't prevent someone from supplying an integer `pos` or `endpos`.